waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -763,14 +763,48 @@
 (defn hard-delete-service-instance
   "Force kill the Kubernetes pod corresponding to the given Waiter Service Instance.
    Does not adjust ReplicaSet replica count; preventing scheduling of a replacement pod must be ensured by the callee.
-   Returns nil on success, but throws on failure."
+   Returns nil on failure, returns the pod object on success."
   [{:keys [api-server-url] :as scheduler} {:keys [service-id] :as instance}]
   (let [pod-url (instance->pod-url api-server-url instance)]
     (try
       (api-request pod-url scheduler :request-method :delete)
       (catch Throwable t
         (log/error t "Error force-killing pod")))))
 
+(defn soft-delete-service-instance
+  [pod-url scheduler body] 
+  (try 
+    (api-request pod-url scheduler :request-method :delete :body body)
+    (catch Throwable t
+      (log/error t "Error soft-killing pod"))))
+
+(defn signal-service-instance
+  "Sends specified signal to the Kubernetes pod corresponding to the given Waiter Service Instance.
+   Returns nil on success, but throws on failure."
+  [{:keys [api-server-url service-id->service-description-fn] {:keys [force-sigterm-secs sigterm-grace-period-secs]} :pod-bypass :as scheduler}
+   {:keys [service-id] :as instance} service signal-type timeout]
+  (let [pod-url (instance->pod-url api-server-url instance)
+        desc (service-id->service-description-fn service-id)
+        bypass-enabled? (sd/service-description-bypass-enabled? desc)
+        ;; We have to add both of these durations together because K8s gracePeriodSeconds countdown is done
+        ;; in parallel with the preStop hook and issuing the SIGTERM signal.
+        ;; https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination
+        force-sigterm-secs (utils/parse-int (get-in desc ["env" "WAITER_CONFIG_BYPASS_FORCE_SIGTERM_SECS"] (str force-sigterm-secs)))
+        sigterm-grace-period-secs (utils/parse-int (get-in desc ["env" "WAITER_CONFIG_BYPASS_SIGTERM_GRACE_PERIOD_SECS"] (str sigterm-grace-period-secs)))
+        total-bypass-grace-period-secs (+ force-sigterm-secs sigterm-grace-period-secs)
+        grace-period-seconds (if bypass-enabled? total-bypass-grace-period-secs 300)]
+
+      (scheduler/log-service-instance instance signal-type :info)
+      (case signal-type
+        
+        ; "soft" delete of the pod (i.e., simply transition the pod to "Terminating" state)
+        :sigterm (soft-delete-service-instance pod-url scheduler (utils/clj->json {:kind "DeleteOptions" :apiVersion "v1" :gracePeriodSeconds (or (if (zero? (Integer/parseInt timeout)) nil (Integer/parseInt timeout)) grace-period-seconds)}))
```

Comment:
The timeout parsing should not happen in this function, it should happen in the handler.